### PR TITLE
Jdb the great expansion

### PR DESCRIPTION
**Pull request by jdblair**
This changes the format of the OSC message so we can send 1 message and update a whole set of beads at once. This appears to fix our performance problems caused by trying to update 480 beads (the individual LEDs on the cross) b/c now they can be sent as 1 update.

Beads now have a "class" value that defines the name of the set. Currently there is rosary, cross and bases. The new format is /bead siib <class> <base> <count> <binary blob>. The base is the starting bead number. The count is the number of beads. The binary blob has 4 bytes per bead in the form r,g,b,brightness.

The pull request also includes the changes to oscled (the C++ controller) and the python simulator to correctly receive the new message format. oscled also includes changes to the configuration file format so we can define the type, or "interface class" of each interface, so we can track the different types of bead updates.

### CODE REVIEW

File: python/mp/color.py [hunk 1]
```diff
@@ -6,6 +6,7 @@ def __init__(self, r=0, g=0, b=0, a=0, name='color'):
         self.g = g  # green
         self.b = b  # blue
         self.a = a  # alpha channel (not yet used)
+        self.brightness = 0xff;  # brightness used by APA102 LED module
```
Comment (jdblair) [inline python/mp/color.py:9]: We're not using this yet, but the APA102 LED module has a 4th value that can be used to control the overall brightness level by controlling the current flowing to all three LEDs

File: python/mp/rosary.py [hunk 2]
```diff
@@ -18,7 +19,7 @@
 class Bead:
     """Bead represents a single rosary bead."""
     def __init__(self, index=0):
-        self.index = index
+        self.index = int(index)
```
Comment (jdblair) [inline python/mp/rosary.py:22]: this was an early experiment to speed things up, b/c I read that python3 iterators are slower if you don't force values to int().

File: python/mp/rosary.py [hunk 3]
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
...
+                                       int(bead.color.g * 0xffff),
+                                       int(bead.color.b * 0xffff),
+                                       int(bead.color.brightness)
+            ))
+
+        msg.add_arg(bytes(payload))
+
+        self.osc_client.send(msg.build())
+
+
```
Comment (jdblair) [inline python/mp/rosary.py:58]: This object contains the context needed to send an OSC message for a set of beads. This context is name (class), the bead_list for the set of beads, and the osc_client object.

File: python/mp/rosary.py [hunk 4]
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
...
-
     """
 
     # Can't decorate with @self.r, so need this here
     dm = DispatcherMapper()
 
     def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
         self.beads = []
+        self.bases = []
+        self.cross = []
```
Comment (jdblair) [inline python/mp/rosary.py]: This is the 2 new sets of beads: bases and cross.

File: python/mp/rosary.py [hunk 5]
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
...
         self.bgcolor = color.Color(0,0,0)
         self.triggers = []
         self.osc_ip = ip
         self.osc_port = port
         self.trigger_id = 0
         self.BEAD_COUNT=60
+        self.BASE_COUNT=9
+        self.CROSS_LED_COUNT=480
         self.run_mainloop = False
```
Comment (jdblair) [inline python/mp/rosary.py:81]: size of two new set of beads

File: python/mp/rosary.py [hunk 6]
```diff
@@ -57,9 +87,29 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
         self.dispatcher = dispatcher
         # Available knobs to turn
         self.knobs = {}
+        self.updater_list = []
```
Comment (jdblair) [inline python/mp/rosary.py:91]: this holds the list of Updater objects that we step through in our mainloop

File: python/mp/rosary.py [hunk 7]
```diff
@@ -57,9 +87,29 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
...
+        self.updater_list.append(Updater(name='base',
+                                         bead_list=self.bases,
+                                         osc_client=self.osc_client))
+
+        for i in range(self.CROSS_LED_COUNT):
+            self.cross.append(Bead(i))
+        self.updater_list.append(Updater(name='cross',
+                                         bead_list=self.cross,
+                                         osc_client=self.osc_client))
+
```
Comment (jdblair) [inline python/mp/rosary.py:113]: create the three sets of beads and the associated Updater

File: python/mp/rosary.py [hunk 8]
```diff
@@ -82,7 +132,9 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
             'even_all': frozenset(self.beads[0:60:2]),
             'even_ring': frozenset(self.beads[4:60:2]),
             'odd_all': frozenset(self.beads[1:60:2]),
-            'odd_ring': frozenset(self.beads[5:60:2])
+            'odd_ring': frozenset(self.beads[5:60:2]),
+            'base': frozenset(self.bases[0:self.BASE_COUNT]),
+            'cross': frozenset(self.cross[0:self.CROSS_LED_COUNT])
         }
```
Comment (jdblair) [inline python/mp/rosary.py:139]: two new set names: base and cross. we're going to want more, but this gets us started.

File: python/mp/rosary.py [hunk 9]
```diff
@@ -122,8 +172,8 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
         # it holds all the other effects.
         self.bin = effects.bin.Bin(self.set_registry['all'], rosary=self)
 
-    def beads_set_bgcolor(self):
-        for bead in self.beads:
+    def beads_set_bgcolor(self, beads):
+        for bead in beads:
             bead.color.set(self.bgcolor)
```
Comment (jdblair) [inline python/mp/rosary.py:178]: reset all the beads in all the bead classes

File: python/mp/rosary.py [hunk 10]
```diff
@@ -29,25 +30,54 @@ def copy_color(self, color):
...
+        msg.add_arg(int(len(self.bead_list)))  # length
+        
+        payload = bytearray()
+        for bead in self.bead_list:
+            payload.extend(struct.pack('!HHHH',
+                                       int(bead.color.r * 0xffff),
+                                       int(bead.color.g * 0xffff),
+                                       int(bead.color.b * 0xffff),
+                                       int(bead.color.brightness)
+            ))
```
Comment (jdblair) [inline python/mp/rosary.py:52]: This is where we pack the binary blob format: r,g,b,brightness in 4 binary bytes. The byte order is big endian (network byte order) b/c that's the tradition, even though it means byte-swapping on both ends of the wire.

File: python/pympsim/simulator.py [hunk 11]
```diff
@@ -41,15 +42,28 @@ def __init__(self, master):
...
+    # def updateBeads(self, beadData):
+    #     for bd in beadData:
+    #         bead = bd[0]
+    #         r = int(bd[1] * 255)
+    #         g = int(bd[2] * 255)
+    #         b = int(bd[3] * 255)
+    #         print(r,g,b)
+    #         tk_rgb = "#%02x%02x%02x" % (r, g, b)
+    #         self.canvas.itemconfig(self.beads[bead], fill=tk_rgb)
+
```
Comment (jdblair) [inline python/pympsim/simulator.py:54]: I left this in here in case we need it again, even though its cruft.

File: python/pympsim/simulator.py [hunk 12]
```diff
@@ -41,15 +42,28 @@ def __init__(self, master):
...
+            num = bd[0]
+            length = bd[1]
+            blob = bd[2]
+            for i in range(0, len(blob), 6):
+                #print(blob[i:i+6])
+                (r, g, b, brightness) = unpack('!HHHH', blob[i:i+6])
+                tk_rgb = "#%02x%02x%02x" % (r >> 8, g >> 8, b >> 8)
+                self.canvas.itemconfig(self.beads[num], fill=tk_rgb)
+                num += 1
+
```
Comment (jdblair) [inline python/pympsim/simulator.py:68]: Decode the binary blob

File: python/pympsim/OSCserver.py [hunk 13]
```diff
@@ -19,6 +19,14 @@ def print_update(addr, args):
...
     queue.put(beadblob)
+    beadblob =[]
+
+def print_bead_rosary(addr, args, base, length, blob):
+    global beadblob
+    queue = args[0]
+
+    beadblob.append([base, length, blob])
+    queue.put(beadblob)
     beadblob = []
```
Comment (jdblair) [inline python/pympsim/OSCserver.py]: push the binary blob on the queue

File: src/opi/PlatformSerialFactory.cpp [hunk 14]
```diff
@@ -8,13 +8,34 @@
...
+    case 2:
+        ser.reset(new PlatformSerial(11, 13));
+        break;
+    case 3:
+        ser.reset(new PlatformSerial(16, 18));
+        break;
+    case 4:
+        ser.reset(new PlatformSerial(8, 10));
+        break;
+    }
```
Comment (jdblair) [inline src/opi/PlatformSerialFactory.cpp:34]: These are the GPIO locations in the new interface board.
See https://oshpark.com/orders/9GaWwwGm

File: src/oscled.json [hunk 15]
```diff
@@ -14,15 +14,24 @@
...
                     "led_count" : 40,
                     "reversed" : false
+                },
+                {
+                    "id": 2,
+                    "class": "base",
+                    "led_base": 0,
+                    "led_count": 8,
+                    "reversed": false
                 }
```
Comment (jdblair) [inline src/oscled.json]: this is an example "base" class. all controllers will have at least one base attached on the third port.

File: src/OSCServer.h [hunk 16]
```diff
@@ -35,6 +35,10 @@ class OSCServer {
 
     OSCServer(std::string ip, std::string port);
 
+    static const int BEAD_ROSARY;
+    static const int BEAD_BASES;
+    static const int BEAD_CROSS;
+
     int bind(std::shared_ptr<IPlatformSerial> const ser, OSCLedConfig::interface_config const &cfg);
```
Comment (jdblair) [inline src/OSCServer.h:42]: whoops, i didn't end up using these.

File: python/mp/rosary.py [hunk 17]
```diff
@@ -231,28 +281,38 @@ def register_written_triggers(self):
...
             
-        bundle = bundle.build()
-        self.osc_client.send(bundle)
+        # msg = osc_message_builder.OscMessageBuilder(address = "/update")
 
+        #bundle.add_content(msg)
+            
+        #bundle = bundle.build()
+        #self.osc_client.send(bundle)
+        self.osc_client.send(msg.build())
```
Comment (GANDAMU) [inline python/mp/rosary.py:313]: If I understand the point of this PR correctly, it's now the job of the Updater classes to call update() and set the colors on the individual beads? What do the changes here do then? Or rather, what does rosary.update() do now?

File: python/mp/rosary.py (hunk 10 above)
Comment (GANDAMU) [reply python/mp/rosary.py:52]: If I understand correctly, we used to send /beadf with a bead ID and a color, and now we're sending /bead with a bead ID start, how many beads from there, and the color?

File: python/mp/rosary.py [hunk 18]
```diff
@@ -57,9 +87,29 @@ def __init__(self, ip="127.0.0.1", port=5005, dispatcher=None, name="rosary"):
...
+        self.updater_list.append(Updater(name='base',
+                                         bead_list=self.bases,
+                                         osc_client=self.osc_client))
+
+        for i in range(self.CROSS_LED_COUNT):
+            self.cross.append(Bead(i))
+        self.updater_list.append(Updater(name='cross',
+                                         bead_list=self.cross,
+                                         osc_client=self.osc_client))
+
```
Comment (GANDAMU) [inline python/mp/rosary.py:112]: Shouldn't the bead class be taking a name arg too now? Unless I'm missing something, just reading this it seems self.bases[0], self.beads[0], and self.cross[0] would all refer to the same bead?

**Review by GANDAMU — CHANGES_REQUESTED** [review]
It all LOOKs good, for the most part, save for my point about the beads. But I can't manage to add any effects.

I'm hoping my one point had something to do with it though.

File: python/mp/rosary.py (hunk 10 above)
Comment (jdblair) [reply python/mp/rosary.py:52]: We're sending the starting bead, the number of beads and then a binary blob that contains the colors for all of those beads. It means we send 3 updates for all the "beads": 1 for the original rosary beads, 1 for the lights in the 9 base structures, and 1 for the 480 LEDs in the cross. It was too expensive to send 60 + 9 + 480 individual updates.

File: python/mp/rosary.py (hunk 18 above)
Comment (jdblair) [reply python/mp/rosary.py:112]: No, Bead should not take a name. self.bases[0], self.beads[0] and self.cross[0] do not refer to the same bead. There are 3 python lists of beads to help us keep them straight, but the effects never see these arrays. The effects get sets of beads, and we can pull those sets from any of the three sets of beads.

Right now if you mix beads from multiple lists in the same set we have a problem b/c they won't sort correctly. I'm "sorting" that out still. Hah. In reality I don't think we will want to combine, say, the cross and the rosary in the same effect so I'm not too concerned.

File: python/mp/rosary.py (hunk 17 above)
Comment (jdblair) [reply python/mp/rosary.py:313]: rosary.update() is cruft now. I should have removed it. Or we can keep it and I can move the code that calls updater there.

File: python/mp/rosary.py (hunk 18 above)
Comment (jdblair) [reply python/mp/rosary.py:112]: To put it another way, here are some examples:

```
# run 3phase_sine_wave on the cross
oscsend localhost 5006 /rosary/add_effect/name/3phase_sine_wave/bead_set/cross
# run throb on the bases
oscsend localhost 5006 /rosary/add_effect/name/throb/bead_set/bases
# run bounce on the rosary
oscsend localhost 5006 /rosary/add_effect/name/bounce/bead_set/rosary
```

**Pull request merged**